src/software/ai/hl/stp/tactic/attacker/attacker_fsm.h

```diff
@@ -103,27 +209,59 @@ struct AttackerFSM
-                                                  ->getEnemyAboutToStealBallRadius()
-                                                  ->value());
-            for (const auto &enemy : event.common.world.enemyTeam().getAllRobots())
-            {
-                if (contains(about_to_steal_danger_zone, enemy.position()))
-                {
-                    return true;
-                }
-            }
-            // otherwise check for shot or pass
-            return event.control_params.pass || event.control_params.shot;
+//        const auto should_kick = [](auto event) {
+//            Point robot_left_side = event.common.robot.position() +
+//                                    Vector::createFromAngle(event.common.robot.orientation()).perpendicular().normalize(
+//                                            ROBOT_MAX_RADIUS_METERS);
+//            Point robot_right_side = event.common.robot.position() + Vector::createFromAngle(
+//                    event.common.robot.orientation()).perpendicular().normalize(ROBOT_MAX_RADIUS_METERS);
+//            Vector stealing_zone = Vector::createFromAngle(event.common.robot.orientation()).normalize(
+//                    event.control_params.attacker_tactic_config->getEnemyAboutToStealBallDistance()->value());
+//            // check for enemy threat
+//            Polygon about_to_steal_danger_zone({
+//                                                       robot_left_side,
+//                                                       robot_left_side + stealing_zone,
+//                                                       robot_right_side + stealing_zone,
+//                                                       robot_right_side
+//                                               });
+//            for (const auto &enemy : event.common.world.enemyTeam().getAllRobots()) {
+//                if (contains(about_to_steal_danger_zone, enemy.position())) {
+//                    return true;
+//                }
+//            }
+//            // otherwise check for shot or pass
+//            return event.control_params.pass || event.control_params.shot;
+//        }
+
+        const auto has_shot = [](auto event) {
+            auto ret =  event.control_params.shot.has_value();
+//            std::cout << "has shot: " << ret << std::endl;
+            return ret;
+        };
+
+        const auto has_pass = [](auto event) {
+            auto ret =  event.control_params.pass.has_value();
+//            std::cout << "has pass: " << ret << std::endl;
+            return ret;
         };
 
         return make_transition_table(
+            // clang-format off
             // src_state + event [guard] / action = dest_state
-            *keep_away_s + update_e[should_kick] / pivot_kick = pivot_kick_s,
-            keep_away_s + update_e[!should_kick] / keep_away,
-            pivot_kick_s + update_e / pivot_kick, pivot_kick_s = X);
+            *keep_away_s + update_e[has_shot] / shoot = shoot_s,
```

Comment:
The transition table is basically a more granular version of the previous one, with different states for passing vs shooting so we can control our transitions for each (and stay committed to a particular one).